### PR TITLE
Documentation/swagger

### PR DESCRIPTION
**Pull request by yosef32**
## A guide to using swagger documentation

	Swagger and swagger ui documentation https://swagger.io/docs
	Swagger ui we will be used outside
	In swagger ui we can test the documentation

	Redoc git repo https://github.com/Redocly/redoc
	Redoc is a ui we using in inside net

	goswagger documentation https://goswagger.io
	go swagger runs on all the files under the swagger folder and creates swagger.json file

	How to use with goswagger
	all commands run in terminal
	install            -  go get -u github.com/go-swagger/go-swagger/cmd/swagger
	run on docker      -  alias swagger="docker run --rm -it -e GOPATH=$HOME/go:/go -v $HOME:$HOME -w $(pwd) quay.io/goswagger/swagger"
	create a json file -  swagger generate spec -o ./swagger/ui/swagger.json --scan-models

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -5,42 +5,10 @@ services:
     build:
       context: .
       dockerfile: Dockerfile
-    environment:
```
Comment (Shahar-Y) [inline docker-compose.yml:8]: Why was it changed?

File: Dockerfile [hunk 2]
```diff
@@ -11,6 +11,6 @@ RUN make build
 #final stage
 FROM scratch
 LABEL Name=api-gateway Version=0.0.1
-COPY --from=builder /go/src/app/api-gateway /api-gateway
```
Comment (Shahar-Y) [inline Dockerfile:14]: why? and by how much has the size of the image increased?

**Review by Shahar-Y — CHANGES_REQUESTED** [review]
- [ ] Deploy in your environment.
 - [ ] Add documentation in the PR on how to use the swagger.
 - [ ] Add changes to CHANGELOG.md

File: Dockerfile (hunk 2 above)
Comment (yosef32) [reply Dockerfile:14]: done

File: server/router.go [hunk 3]
```diff
@@ -155,6 +156,32 @@ func NewRouter(logger *logrus.Logger) (*gin.Engine, []*grpc.ClientConn) {
 	// Health Check route.
 	apiRoutesGroup.GET("/healthcheck", health.healthCheck)
 
+	// handler for swagger documentation
+	apiRoutesGroup.StaticFile("/swagger", viper.GetString(configSwaggerPathFile) + "/swagger.json")
+
+	if viper.GetString(configDeployment) == "prodaction" {
```
Comment (Shahar-Y) [inline server/router.go:162]: 1. Was it tested with and without the environment variable? 
2. Change `prodaction` to `production`.
3. the string `production` should be a constant.

File: server/server.go [hunk 4]
```diff
@@ -84,8 +86,10 @@ func init() {
 	viper.SetDefault(configExternalShareName, "שיתוף חיצוני")
 	viper.SetDefault(configMyExternalSharesName, "השיתופים החיצוניים שלי")
 	viper.SetDefault(configVipService, "http://localhost:8094")
+	viper.SetDefault(configDeployment, "integration")
```
Comment (Shahar-Y) [inline server/server.go:89]: Is it good that the default is integration which allows more options for the user? 
On second thought, maybe it is better to have an environment variable which says "allow swagger?" and the default would be false.

File: swagger/file.swagger.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+import (
+	file "github.com/meateam/api-gateway/file"
+)
+
+// swagger:route GET /files files listFiles
+//
+// List of files
+//
+// This will return all files according to the requested folder
```
Comment (Shahar-Y) [inline swagger/file.swagger.go:21]: remove `will`, ass `s` to returns on all occurrences

**Review by Shahar-Y — CHANGES_REQUESTED** [review]
Upload to your environment, and add to the changelog

File: docker-compose.yml (hunk 1 above)
Comment (yosef32) [reply docker-compose.yml:8]: done

File: server/router.go (hunk 3 above)
Comment (yosef32) [reply server/router.go:162]: 1. yes
2. done
3. done

File: server/server.go (hunk 4 above)
Comment (yosef32) [reply server/server.go:89]: Yes because the development environment is integration

File: swagger/file.swagger.go (hunk 5 above)
Comment (yosef32) [reply swagger/file.swagger.go:21]: done

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,13 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [v2.2.0] - 2020-11-9
+
+### Added
+
+- Swagger documentation 
+  - Naw can go to /api/docs and see the documentation
```
Comment (Shahar-Y) [inline CHANGELOG.md:13]:
```suggestion
  - Now you can go to /api/docs and see the documentation 
```

File: server/server.go (hunk 4 above)
Comment (Shahar-Y) [reply server/server.go:89]: But you want the default to be safer, without an option to make changes in prod

File: swagger/permit.swagger.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	Permits ppb.CreatePermitResponse
+}
+
+// rg.PATCH(fmt.Sprintf("/permits/:%s", ParamReqID), checkStatusScope, r.UpdateStatus)
+
+// swagger:route PATCH /files/{id}/permits files patchpermits
+//
+// Update file permits
+//
+// This will update the permits to the file
```
Comment (Shahar-Y) [inline swagger/permit.swagger.go:98]:
```suggestion
// This updates the permits to the file
```

**Review by Shahar-Y — CHANGES_REQUESTED** [review]

File: server/server.go (hunk 4 above)
Comment (yosef32) [reply server/server.go:89]: done

File: server/router.go [hunk 8]
```diff
@@ -155,6 +158,32 @@ func NewRouter(logger *logrus.Logger) (*gin.Engine, []*grpc.ClientConn) {
...
 	apiRoutesGroup.GET("/healthcheck", health.healthCheck)
 
+	// handler for swagger documentation
+	apiRoutesGroup.StaticFile("/swagger", viper.GetString(configSwaggerPathFile) + "/swagger.json")
+
+	if viper.GetString(configDeployment) == deploymentProduction {
+		opts := middleware.RedocOpts{
+			SpecURL:  "/api/swagger",
+			BasePath: "/api",
+			RedocURL:  "/api/redoc",
```
Comment (Shahar-Y) [inline server/router.go:168]: Wait, so production allows redoc?
I believe that the best solution is to use the basic swagger by default, and if the environment is "integration" - then add redoc

**Review by Shahar-Y — CHANGES_REQUESTED** [review]

File: server/router.go (hunk 8 above)
Comment (yosef32) [reply server/router.go:168]: done

**Review by Shahar-Y — APPROVED** [review]

**Comment by Shahar-Y** [comment]
@yosef32 make sure you resolve the conflicts in the changelog properly

**Review by Shahar-Y — DISMISSED** [review]

**Review by Shahar-Y — APPROVED** [review]

**Pull request merged**